src/analysis/processing/qgsalgorithmdissolve.cpp

```diff
@@ -212,9 +212,27 @@ QgsDissolveAlgorithm *QgsDissolveAlgorithm::createInstance() const
 
 QVariantMap QgsDissolveAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
 {
-  return processCollection( parameters, context, feedback, []( const QVector< QgsGeometry > &parts )->QgsGeometry
+  return processCollection( parameters, context, feedback, [ & ]( const QVector< QgsGeometry > &parts )->QgsGeometry
   {
-    return QgsGeometry::unaryUnion( parts );
+    QgsGeometry result( QgsGeometry::unaryUnion( parts ) );
+    // Geos may fail in some cases, let's try a slower but safer approach
+    // See: https://issues.qgis.org/issues/20591 - Dissolve tool failing to produce outputs
+    if ( ! result.lastError().isEmpty() && parts.count() >  2 )
+    {
+      QgsDebugMsg( QStringLiteral( "GEOS exception, taking the slower route ..." ) );
```

Comment:
Do we really need this debug msg within the algorithm?

Reply:
It's debatable, of course.  

I think that this solution is a bit of a workaround: the real solution (if it does exist) should be in GEOS, this route can be significantly slower if you have many features, so I think it's important to know that we took the slow route, in case someone is checking why it is so slow at least there is something in the logs.

But really, is it so important if there is a debug message or not?

Reply:
I'd say push that info into the algorithm dialog, or simply remove it altogether.

It's certainly not a blocker for pushing this fix onto master, it's just that QGIS can IMHO be quite noisy already when it comes to debugging msgs.

Reply:
I agree on the noisy part, there are a lot of messages that could go into a msglevel, I'll see if it can be pushed to the dialog, but for now I'm a bit in a hurry at least for the backport.